### ISSUE TITLE
Indicate expired tag life in deployment end dates in R and on Motus.org

### ISSUE THREAD

**Issue by j-sayers** [MotusDev/Motus-TO-DO#947]
Project 20 recently contacted us regarding some tags whose deployments appear to be missing in the data in the R package. The tags in question have had two deployments. The earlier ones show up as expected, but there are no deployments associated with detections for the later deployments.

The tags have correct, active, deployment information in the Motus site and in the database. But digging in to the tagDeps table in the R package, they have **deployment end dates that are earlier than the start date.** Along with this they have a tsEndCode of 3 (?). Calling forceMeta = T and metadata(motus,file) have both failed to update the tag deployment metadata.

![image](https://user-images.githubusercontent.com/55001144/114452654-b47e8900-9ba6-11eb-8c9c-e0bb68299b84.png)

**Comment by cjardine-bsc** [comment]
The date ends are being calculated based on the tag lifespan. They consider previous deployments. So the problem here is bad metadata. 

The first tag for example:

34071 has an expected lifespan of 92 days. 

It's first deployment is from April 2019-October 2019 
Then it has a second deployment starting Oct 2019. But by that time the tag has used 188 days of lifespan of it's expected 92 days. The tag is dead, and has -96 days of life left. 

Obviously I need to make a change that defaults negative lifespans to 0. But in the meantime they need to look at their deployment metadata for these tags and fix it up.

**Comment by jsayers-bsc** [comment]
Good know...

I take it tsEndCode == 3 indicates it's a calculated field?

Aside from not having those negative, what about just reflecting what's on the motus site and the database -- if the end date is NULL in the db then it's displayed as NULL? 

Though actually, indicating that a deployment with no end date is assumed to be dead is helpful, and would be on the Motus site too. So instead of say, deployments from 5 years ago still saying NULL as their end date, it should reflect the fact that that tag is assumed to be dead.

**Comment by jsayers-bsc** [comment]
Below is a screenshot of tag deployments from 2014, where the deployment end date is given as "?". This is how deployment end dates are displayed unless one has manually been set. We don't want people to manually set them unless the tag has been recovered, but leaving it open-ended is a bit misleading (and is in conflict with how it's currrently dealt with in the R package, which needs some tweaking itself).

![image](https://user-images.githubusercontent.com/55001144/114707246-fbc85f00-9cf7-11eb-960e-d83c352d1244.png)

It would be helpful to change how deployment date is displayed when the tag in question has expired in both the Motus R package and Motus.org to indicate when an expired tag impacts a deployment.

Something like "tag battery expired" would work, though that would add a second line in that cell. Maybe there's something better to use instead?

**Comment by cjardine-bsc** [comment]
We have some text fragments to update. But the tag deployment page now shows this data in the recent push to live. 

![image](https://user-images.githubusercontent.com/22172758/114766097-d731ae80-9d1a-11eb-8a93-427f8511a5a9.png)

**Comment by jsayers-bsc** [comment]
To clarify, I was actually referring to the deployment date in the manage tags page, which currently shows a question mark for end date even if the tag End of Life date is in the past. If we auto-populate this field based on the estimated battery life that would clarify this whole process for many people 

At the very least, I think we should indicate when the end of life date has passed and get rid of the question mark in those cases. I'm not sure what that should be replaced with though. Something like "Tag battery presumed dead" is descriptive enough, but perhaps too wordy.

There may even be a case for providing this calculated end of life countdown in the Manage Tags page for all deployments. Deployment period and battery life are central to how tag finder matches raw data up with tags, but I don't think most people get the importance of this. Making the battery life side of this more conspicuous may help get this across, and maybe even encourage better metadata management in general.

**Comment by cjardine-bsc** [comment]
Lets talk about this on Thursday. It's easy enough to pass the calculated end dates if that's what we decide we want.

**Comment by pmorrill** [comment]
Agreed. And it is also possible to post expired end dates on the mgmt tools, once relevant.

**Comment by cjardine-bsc** [comment]
I've included the end dates in the views. That updated things in live: https://motus.org/data/project/tags

Sandbox didn't auto-update: https://sandbox.motus.org/data/project/tags?projectID=1

**Comment by jsayers-bsc** [comment]
Can we make format this something like gray and italics so it’s more obviously no user-entered?

Would be nice to include this actual calculated date in the Edit Tags page too (where the calculated buffers are displayed).

Do the deployment summaries in the Edit Tags page come from a different source? The dates aren’t present there…

**Comment by jsayers-bsc** [comment]
I'm being constantly thrown by this since it looks like a regular deployment date. I haven't heard from others yet, but I suspect we'll be getting equally confused emails.

Until we have the cosmetics in place to make it clear that this is an _estimated_ date, maybe we should keep the question mark. @cjardine-bsc how simple is it to reverse the change?

**Comment by cjardine-bsc** [comment]
Careful what you wish for ;-) 

Reverted in live. 

@pmorrill  I'll have to add a field the the vwBase_tag_dep and vwData_tag_dep to flag calculated vs hard dates and you can format accordingly. 

I've done that in sandbox. The field is calc_date true means the date is calculated, false is a hard date.

**Comment by pmorrill** [comment]
If a deployment is still active, we want to continue to display the question mark to illustrate that. Therefore I find it confusing to have the dtEnd field automatically populated with a 'calculated end date' in all cases where dt_end is null. In fact, I only want to use the calculated end date if the dtEnd is null, **and the calculated end date is in the past**.

(Otherwise I want to continue to display a question mark.)

Can you set it so that the calculated field is passed as a new field (dtCalculatedEnd), and the calc_date field now becomes calc_date_valid (set only if the dtCalculatedEnd field is in the past).

**Comment by cjardine-bsc** [comment]
Both dtEnd and dtEndCalc  are now in all the views in live and sandbox. Have fun!

**Comment by pmorrill** [comment]
Have now adjusted beta10 and master tags tools to use the new dtEndCalc field in several spots.

**Comment by j-sayers** [comment]
This has been implimented
